File: frontend/src/lib/scheduling.ts
import { apiRequest } from "@/lib/auth";

export type CalendarProvider = "google" | "outlook" | "calcom" | "external_link";
export type BookingStatus =
  | "pending"
  | "confirmed"
  | "cancelled"
  | "rescheduled"
  | "completed"
  | "no_show";

export type CalendarConnectionSummary = {
  id: string;
  workspace_id: string;
  user_id: string | null;
  provider: CalendarProvider;
  status: string;
  external_account_id: string | null;
  external_account_email: string | null;
  metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type MeetingTypeSummary = {
  id: string;
  workspace_id: string;
  title: string;
  slug: string;
  description: string | null;
  duration_minutes: number;
  location_type: "google_meet" | "external_url" | "manual" | "phone" | "in_person";
  assigned_user_id: string | null;
  assignment_mode: "specific" | "round_robin";
  provider_preference: CalendarProvider | null;
  external_location_url: string | null;
  manual_location_text: string | null;
  booking_link: string;
  availability_rules: Record<string, unknown> | null;
  metadata: Record<string, unknown> | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type AvailabilityRule = {
  weekday: number;
  start_time: string;
  end_time: string;
  is_enabled: boolean;
};

export type AvailabilitySettings = {
  timezone: string;
  buffer_before_minutes: number;
  buffer_after_minutes: number;
  duration_options: number[];
  max_bookings_per_day: number;
  minimum_notice_minutes: number;
  future_booking_window_days: number;
  fallback_owner_user_id: string | null;
  reminder_in_app_enabled: boolean;
};

export type AvailabilityResponse = {
  workspace_id: string;
  meeting_type_id: string | null;
  user_id: string | null;
  rules: AvailabilityRule[];
  settings: AvailabilitySettings;
};

export type BlackoutDateSummary = {
  id: string;
  workspace_id: string;
  meeting_type_id: string | null;
  user_id: string | null;
  title: string | null;
  reason: string | null;
  start_time_utc: string;
  end_time_utc: string;
  timezone: string;
  created_at: string;
};

export type AvailableSlot = {
  start_time_utc: string;
  end_time_utc: string;
  display_time: string;
  timezone: string;
  provider: string;
  assigned_user_id: string | null;
};

export type BookingSummary = {
  id: string;
  workspace_id: string;
  lead_id: string | null;
  chat_session_id: string | null;
  meeting_type_id: string;
  meeting_type_title: string;
  assigned_user_id: string | null;
  assigned_user_name: string | null;
  visitor_name: string;
  visitor_email: string;
  visitor_phone: string | null;
  start_time_utc: string;
  end_time_utc: string;
  timezone: string;
  status: BookingStatus;
  provider: string;
  external_event_id: string | null;
  meeting_link: string | null;
  management_url: string;
  public_reschedule_url: string;
  public_cancel_url: string;
  attendees: Array<{
    id: string;
    name: string;
    email: string;
    phone: string | null;
    timezone: string | null;
    is_primary: boolean;
  }>;
  metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type PublicBookingWorkspaceResponse = {
  workspace_id: string;
  workspace_name: string;
  workspace_slug: string;
  admin_timezone: string;
  meeting_types: MeetingTypeSummary[];
};

export async function fetchCalendarStatus(workspaceId: string) {
  return apiRequest<{ workspace_id: string; items: CalendarConnectionSummary[] }>(
    `/api/v1/calendar/status?workspace_id=${encodeURIComponent(workspaceId)}`,
  );
}

export async function connectCalendar(payload: {
  workspace_id: string;
  provider: CalendarProvider;
  user_id?: string | null;
  api_key?: string;
  external_booking_url?: string;
  external_account_email?: string;
}) {
  return apiRequest<CalendarConnectionSummary>("/api/v1/calendar/connect", {
    method: "POST",
    json: payload,
  });
}

export async function disconnectCalendar(payload: {
  workspace_id: string;
  provider: CalendarProvider;
  user_id?: string | null;
}) {
  return apiRequest<{ message: string }>("/api/v1/calendar/disconnect", {
    method: "DELETE",
    json: payload,
  });
}

export async function listMeetingTypes(workspaceId: string) {
  return apiRequest<{ items: MeetingTypeSummary[] }>(
    `/api/v1/meeting-types/list?workspace_id=${encodeURIComponent(workspaceId)}`,
  );
}

export async function createMeetingType(payload: Record<string, unknown>) {
  return apiRequest<MeetingTypeSummary>("/api/v1/meeting-types/create", {
    method: "POST",
    json: payload,
  });
}

export async function updateMeetingType(id: string, payload: Record<string, unknown>) {
  return apiRequest<MeetingTypeSummary>(`/api/v1/meeting-types/${encodeURIComponent(id)}`, {
    method: "PUT",
    json: payload,
  });
}

export async function deleteMeetingType(workspaceId: string, id: string) {
  return apiRequest<{ message: string }>(
    `/api/v1/meeting-types/${encodeURIComponent(id)}?workspace_id=${encodeURIComponent(workspaceId)}`,
    { method: "DELETE" },
  );
}

export async function fetchAvailability(workspaceId: string) {
  return apiRequest<AvailabilityResponse>(
    `/api/v1/availability/get?workspace_id=${encodeURIComponent(workspaceId)}`,
  );
}

export async function setAvailability(payload: Record<string, unknown>) {
  return apiRequest<AvailabilityResponse>("/api/v1/availability/set", {
    method: "POST",
    json: payload,
  });
}

export async function listBlackouts(workspaceId: string) {
  return apiRequest<BlackoutDateSummary[]>(
    `/api/v1/availability/blackout/list?workspace_id=${encodeURIComponent(workspaceId)}`,
  );
}

export async function createBlackout(payload: Record<string, unknown>) {
  return apiRequest<BlackoutDateSummary>("/api/v1/availability/blackout", {
    method: "POST",
    json: payload,
  });
}

export async function deleteBlackout(workspaceId: string, blackoutId: string) {
  return apiRequest<{ message: string }>(
    `/api/v1/availability/blackout/${encodeURIComponent(blackoutId)}?workspace_id=${encodeURIComponent(workspaceId)}`,
    { method: "DELETE" },
  );
}

export async function fetchBookingSlots(params: {
  workspace_id: string;
  meeting_type_id: string;
  timezone: string;
  start_date?: string;
  end_date?: string;
  lead_id?: string | null;
  chat_session_id?: string | null;
  reschedule_booking_id?: string | null;
}) {
  const query = new URLSearchParams({
    workspace_id: params.workspace_id,
    meeting_type_id: params.meeting_type_id,
    timezone: params.timezone,
  });
  if (params.start_date) query.set("start_date", params.start_date);
  if (params.end_date) query.set("end_date", params.end_date);
  if (params.lead_id) query.set("lead_id", params.lead_id);
  if (params.chat_session_id) query.set("chat_session_id", params.chat_session_id);
  if (params.reschedule_booking_id) query.set("reschedule_booking_id", params.reschedule_booking_id);
  return apiRequest<{ workspace_id: string; meeting_type_id: string; timezone: string; slots: AvailableSlot[] }>(
    `/api/v1/booking/slots?${query.toString()}`,
  );
}

export async function createBooking(payload: Record<string, unknown>) {
  return apiRequest<BookingSummary>("/api/v1/booking/create", {
    method: "POST",
    json: payload,
  });
}

export async function rescheduleBooking(payload: Record<string, unknown>) {
  return apiRequest<BookingSummary>("/api/v1/booking/reschedule", {
    method: "POST",
    json: payload,
  });
}

export async function cancelBooking(payload: Record<string, unknown>) {
  return apiRequest<{ message: string }>("/api/v1/booking/cancel", {
    method: "POST",
    json: payload,
  });
}

export async function listBookings(params: {
  workspace_id: string;
  status?: string;
  date_from?: string;
  date_to?: string;
  assigned_user_id?: string;
}) {
  const query = new URLSearchParams({ workspace_id: params.workspace_id });
  if (params.status) query.set("status", params.status);
  if (params.date_from) query.set("date_from", params.date_from);
  if (params.date_to) query.set("date_to", params.date_to);
  if (params.assigned_user_id) query.set("assigned_user_id", params.assigned_user_id);
  return apiRequest<{ items: BookingSummary[]; total: number }>(`/api/v1/booking/list?${query.toString()}`);
}

export async function fetchBooking(workspaceId: string, bookingId: string, token?: string) {
  const query = new URLSearchParams({ workspace_id: workspaceId });
  if (token) query.set("token", token);
  return apiRequest<BookingSummary>(`/api/v1/booking/${encodeURIComponent(bookingId)}?${query.toString()}`);
}

export async function fetchBookingByToken(token: string) {
  return apiRequest<BookingSummary>(`/api/v1/booking/manage/${encodeURIComponent(token)}`);
}

export async function fetchPublicBookingWorkspace(workspaceSlug: string) {
  return apiRequest<PublicBookingWorkspaceResponse>(
    `/api/v1/booking/public/${encodeURIComponent(workspaceSlug)}`,
  );
}

File: backend/app/schemas/scheduling.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.core.input_validator import sanitize_text, validate_webhook_url

CalendarProviderName = Literal["google", "outlook", "calcom", "external_link"]
BookingStatus = Literal["pending", "confirmed", "cancelled", "rescheduled", "completed", "no_show"]
LocationType = Literal["google_meet", "external_url", "manual", "phone", "in_person"]
AssignmentMode = Literal["specific", "round_robin"]


class CalendarProviderDescriptor(BaseModel):
    provider: CalendarProviderName
    label: str
    auth_type: str
    supports_busy_lookup: bool
    supports_event_creation: bool
    supports_video_link: bool
    supports_oauth: bool


class CalendarConnectRequest(BaseModel):
    workspace_id: uuid.UUID
    provider: CalendarProviderName
    user_id: uuid.UUID | None = None
    access_token: str | None = Field(default=None, max_length=5000)
    refresh_token: str | None = Field(default=None, max_length=5000)
    api_key: str | None = Field(default=None, max_length=5000)
    external_account_id: str | None = Field(default=None, max_length=255)
    external_account_email: str | None = Field(default=None, max_length=255)
    external_booking_url: str | None = Field(default=None, max_length=2000)
    metadata: dict | None = None


class CalendarDisconnectRequest(BaseModel):
    workspace_id: uuid.UUID
    provider: CalendarProviderName
    user_id: uuid.UUID | None = None


class CalendarConnectionSummary(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    user_id: uuid.UUID | None
    provider: CalendarProviderName
    status: str
    external_account_id: str | None
    external_account_email: str | None
    metadata: dict | None
    created_at: datetime
    updated_at: datetime


class CalendarStatusResponse(BaseModel):
    workspace_id: uuid.UUID
    items: list[CalendarConnectionSummary]


class MeetingTypeCreateRequest(BaseModel):
    workspace_id: uuid.UUID
    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=4000)
    duration_minutes: int = Field(ge=15, le=60)
    location_type: LocationType
    assigned_user_id: uuid.UUID | None = None
    assignment_mode: AssignmentMode = "specific"
    provider_preference: CalendarProviderName | None = None
    external_location_url: str | None = Field(default=None, max_length=2000)
    manual_location_text: str | None = Field(default=None, max_length=1000)
    availability_rules: dict | None = None
    metadata: dict | None = None

    @field_validator("title", "description", "manual_location_text", mode="before")
    @classmethod
    def sanitize_text_fields(cls, value: str | None) -> str | None:
        return sanitize_text(value, max_length=4000)


class MeetingTypeUpdateRequest(MeetingTypeCreateRequest):
    pass


class MeetingTypeSummary(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    title: str
    slug: str
    description: str | None
    duration_minutes: int
    location_type: LocationType
    assigned_user_id: uuid.UUID | None
    assignment_mode: AssignmentMode
    provider_preference: CalendarProviderName | None
    external_location_url: str | None
    manual_location_text: str | None
    booking_link: str
    availability_rules: dict | None
    metadata: dict | None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class MeetingTypeListResponse(BaseModel):
    items: list[MeetingTypeSummary]


class WeeklyAvailabilityRuleInput(BaseModel):
    weekday: int = Field(ge=0, le=6)
    start_time: str = Field(pattern=r"^\d{2}:\d{2}$")
    end_time: str = Field(pattern=r"^\d{2}:\d{2}$")
    is_enabled: bool = True


class AvailabilitySettingsInput(BaseModel):
    timezone: str = Field(min_length=1, max_length=100)
    buffer_before_minutes: int = Field(default=0, ge=0, le=180)
    buffer_after_minutes: int = Field(default=0, ge=0, le=180)
    duration_options: list[int] = Field(default_factory=lambda: [15, 30, 45, 60])
    max_bookings_per_day: int = Field(default=20, ge=1, le=200)
    minimum_notice_minutes: int = Field(default=60, ge=0, le=60 * 24 * 14)
    future_booking_window_days: int = Field(default=30, ge=1, le=365)
    fallback_owner_user_id: uuid.UUID | None = None
    reminder_in_app_enabled: bool = False


class AvailabilitySetRequest(BaseModel):
    workspace_id: uuid.UUID
    meeting_type_id: uuid.UUID | None = None
    user_id: uuid.UUID | None = None
    rules: list[WeeklyAvailabilityRuleInput]
    settings: AvailabilitySettingsInput


class AvailabilityResponse(BaseModel):
    workspace_id: uuid.UUID
    meeting_type_id: uuid.UUID | None = None
    user_id: uuid.UUID | None = None
    rules: list[WeeklyAvailabilityRuleInput]
    settings: AvailabilitySettingsInput


class BlackoutDateCreateRequest(BaseModel):
    workspace_id: uuid.UUID
    meeting_type_id: uuid.UUID | None = None
    user_id: uuid.UUID | None = None
    title: str | None = Field(default=None, max_length=255)
    reason: str | None = Field(default=None, max_length=2000)
    start_time: datetime
    end_time: datetime
    timezone: str = Field(min_length=1, max_length=100)


class BlackoutDateSummary(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    meeting_type_id: uuid.UUID | None
    user_id: uuid.UUID | None
    title: str | None
    reason: str | None
    start_time_utc: datetime
    end_time_utc: datetime
    timezone: str
    created_at: datetime


class AvailableSlot(BaseModel):
    start_time_utc: datetime
    end_time_utc: datetime
    display_time: str
    timezone: str
    provider: CalendarProviderName | Literal["internal"]
    assigned_user_id: uuid.UUID | None


class BookingSlotsRequest(BaseModel):
    workspace_id: uuid.UUID
    meeting_type_id: uuid.UUID
    start_date: datetime | None = None
    end_date: datetime | None = None
    timezone: str = Field(min_length=1, max_length=100)
    lead_id: uuid.UUID | None = None
    chat_session_id: uuid.UUID | None = None
    reschedule_booking_id: uuid.UUID | None = None


class BookingSlotsResponse(BaseModel):
    workspace_id: uuid.UUID
    meeting_type_id: uuid.UUID
    timezone: str
    slots: list[AvailableSlot]


class BookingCreateRequest(BaseModel):
    workspace_id: uuid.UUID
    meeting_type_id: uuid.UUID
    lead_id: uuid.UUID | None = None
    chat_session_id: uuid.UUID | None = None
    visitor_name: str = Field(min_length=1, max_length=255)
    visitor_email: str = Field(min_length=5, max_length=255)
    visitor_phone: str | None = Field(default=None, max_length=50)
    start_time_utc: datetime
    timezone: str = Field(min_length=1, max_length=100)
    preferred_location_type: LocationType | None = None
    preferred_date_note: str | None = Field(default=None, max_length=1000)

    @field_validator("visitor_name", "preferred_date_note", mode="before")
    @classmethod
    def sanitize_booking_fields(cls, value: str | None) -> str | None:
        return sanitize_text(value, max_length=1000)


class BookingRescheduleRequest(BaseModel):
    workspace_id: uuid.UUID
    booking_id: uuid.UUID | None = None
    token: str | None = Field(default=None, max_length=255)
    start_time_utc: datetime
    timezone: str = Field(min_length=1, max_length=100)


class BookingCancelRequest(BaseModel):
    workspace_id: uuid.UUID
    booking_id: uuid.UUID | None = None
    token: str | None = Field(default=None, max_length=255)
    reason: str | None = Field(default=None, max_length=1000)


class BookingListRequest(BaseModel):
    workspace_id: uuid.UUID
    status: BookingStatus | None = None
    date_from: datetime | None = None
    date_to: datetime | None = None
    assigned_user_id: uuid.UUID | None = None


class BookingAttendeeSummary(BaseModel):
    id: uuid.UUID
    name: str
    email: str
    phone: str | None
    timezone: str | None
    is_primary: bool


class BookingSummary(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    lead_id: uuid.UUID | None
    chat_session_id: uuid.UUID | None
    meeting_type_id: uuid.UUID
    meeting_type_title: str
    assigned_user_id: uuid.UUID | None
    assigned_user_name: str | None
    visitor_name: str
    visitor_email: str
    visitor_phone: str | None
    start_time_utc: datetime
    end_time_utc: datetime
    timezone: str
    status: BookingStatus
    provider: str
    external_event_id: str | None
    meeting_link: str | None
    management_url: str
    public_reschedule_url: str
    public_cancel_url: str
    attendees: list[BookingAttendeeSummary]
    metadata: dict | None
    created_at: datetime
    updated_at: datetime


class BookingListResponse(BaseModel):
    items: list[BookingSummary]
    total: int


class PublicBookingWorkspaceResponse(BaseModel):
    workspace_id: uuid.UUID
    workspace_name: str
    workspace_slug: str
    admin_timezone: str
    meeting_types: list[MeetingTypeSummary]


class SchedulingIntentResponse(BaseModel):
    detected: bool
    reason: str | None = None
    should_prompt_for_booking: bool = False
    suggested_meeting_type_id: uuid.UUID | None = None

